src/CoreWCF.Primitives/src/CoreWCF/Runtime/SimpleAsyncLock.cs

```diff
@@ -0,0 +1,61 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace CoreWCF.Runtime
+{
+    /// <summary>
+    /// This class provides a locking mechanism compatible with task async patterns.
+    /// </summary>
+    /// <remarks>
+    /// Unlike the <see cref="AsyncLock"/> it does not have the complexity of lock
+    /// tracking and it ensures SemaphoreSlim does not leak memory or handles by hiding
+    /// the access to certain members which would create resources that require disposal.
+    /// </remarks>
+    internal sealed class SimpleAsyncLock
+    {
+        private readonly SemaphoreSlim _semaphore = new SemaphoreSlim(1);
+
+        public async Task<IDisposable> TakeLockAsync(TimeSpan timeout, CancellationToken cancellationToken = default)
+        {
+            await _semaphore.WaitAsync(timeout, cancellationToken);
+            return new SafeSemaphoreRelease(this);
+        }
+
+        public async Task<IDisposable> TakeLockAsync(CancellationToken cancellationToken = default)
+        {
+            await _semaphore.WaitAsync(cancellationToken);
+            return new SafeSemaphoreRelease(this);
+        }
+
+        public IDisposable TakeLock(CancellationToken cancellationToken = default)
+        {
+            _semaphore.Wait(cancellationToken);
+            return new SafeSemaphoreRelease(this);
+        }
+
+        public IDisposable TakeLock(TimeSpan timeout, CancellationToken cancellationToken = default)
+        {
+            _semaphore.Wait(timeout, cancellationToken);
+            return new SafeSemaphoreRelease(this);
+        }
+
+        private readonly struct SafeSemaphoreRelease : IDisposable
+        {
+            private readonly SimpleAsyncLock _asyncLock;
+
+            public SafeSemaphoreRelease(SimpleAsyncLock asyncLock)
+            {
+                _asyncLock = asyncLock;
+            }
+
+            public void Dispose()
+            {
+                _asyncLock._semaphore.Release();
```

Comment:
This is vulnerable to a double release. Add a null check and after releasing the semaphore, null it out.
```suggestion
                _asyncLock?._semaphore.Release();
                _asyncLock = null;
```
Maybe add an Fx.Assert that _asyncLock is not null to catch this. Fx.Assert is DEBUG conditional so has zero overhead in release builds. I'm not adding Fx.Assert to the suggestion as it requires a `using System.Runtime;` at the top too.

Reply:
A double release would imply a misusage of the class because you would really have to initiate a double dispose. It is not finalizable and it is expected to the caller to release the lock via dispose. And if we're concerned about double release, we would also need to be concerned about race conditions on the release. There could be still a double release in this case. 

Even though I think it might be a YAGNI, on thread concurrency related code like this is is usually better to be safe then sorry. Murphys Law is calling 😁 Therefore I would rather recommend to go for a Interlocked.Exchange approach here to be really safe for concurrency and misusages: 

```cs
var asyncLock = Interlocked.Exchange(ref _asyncLock, null);
Fx.Assert(asyncLock != null, "Double unlock, this must be avoided!")
asyncLock?._semaphore.Release();
```